Allow user to edit Extension Requests before approval

ahsrah7 opened the pull request:
<!-- Date Here in dd mmm yyyy-->
Date: 17-10-2024
<!--Developer Name Here-->
Developer Name: Harshavardhan

---

## Issue Ticket Number
<!--Issue ticket this PR closes-->
- closes https://github.com/Real-Dev-Squad/website-backend/issues/2195
## Description

<!--Description of the changes made in this PR-->
- Allows user to edit an extension request if  it is not yet approved.
### Documentation Updated?

- [ ] Yes
- [x] No

<!--Additional notes about documentation update if applicable-->

### Under Feature Flag

- [x] Yes
- [ ] No

<!--Indicate if changes are under a feature flag-->

### Database Changes

- [ ] Yes
- [x] No

<!--Notes on any database changes-->

### Breaking Changes

- [ ] Yes
- [x] No

<!--Notes on breaking changes or related issue tickets if applicable-->

### Development Tested?

- [x] Yes
- [ ] No

<!--Confirmation of local testing during development-->

## Screenshots
<details>
<summary>Screenshot 1</summary>

<!-- Attach your screenshots here👇-->

</details>

https://github.com/user-attachments/assets/93e6636a-a572-4cbb-a50b-722e3ed52fb8


<!--Attach or link to relevant screenshots or visual aids-->

## Test Coverage
<details>
<summary>Screenshot 1</summary>

<!-- Attach your screenshots here👇-->
![image](https://github.com/user-attachments/assets/f07b3556-2822-4b42-ba0c-ac043c25aaa5)
![image](https://github.com/user-attachments/assets/e71bb86f-7299-4cd2-b531-df823d97d812)
![image](https://github.com/user-attachments/assets/1bdd3d36-6333-44b9-8073-39d648813d94)
![image](https://github.com/user-attachments/assets/8c75bce8-1824-4a4a-9409-1a361d1722ec)
![image](https://github.com/user-attachments/assets/5664fbad-81e8-4b39-abb3-5bc073df243d)
![image](https://github.com/user-attachments/assets/dcaf9469-30b0-427a-87bf-d8aa212b1ad4)




</details>

<!--Attach Details on test coverage and outcomes-->

## Additional Notes

<!--Any additional notes, considerations or explanations for reviewers-->

pankajjs commented on routes/extensionRequests.js:
```diff
@@ -18,7 +19,7 @@ router.get("/:id", authenticate, authorizeRoles([SUPERUSER, APPOWNER]), extensio
   "/:id",
   authenticate,
-  authorizeRoles([SUPERUSER, APPOWNER]),
+  skipIfDev(authorizeRoles([SUPERUSER, APPOWNER])),
```
1. skipIfDev does not describe what it does by its name.

ahsrah7 replied:
updated the name. Please have a look

pankajjs commented:
There are no tests for disabled feature flag.

ahsrah7 commented:
> 1. There are no tests for disabled feature flag.
> 2. Could you add the test coverage for you changes?

- will the tests for the feature flag not become the tests once feature flag is removed?
- I have added the relevant test cases but  if you have any specific cases please mention

pankajjs commented:
> > 1. There are no tests for disabled feature flag.
> > 2. Could you add the test coverage for you changes?
> 
> * will the tests for the feature flag not become the tests once feature flag is removed?
> * I have added the relevant test cases but  if you have any specific cases please mention

I don't see 

> > 1. There are no tests for disabled feature flag.
> > 2. Could you add the test coverage for you changes?
> 
> * will the tests for the feature flag not become the tests once feature flag is removed?
> * I have added the relevant test cases but  if you have any specific cases please mention

@ahsrah7 I don't see any test when dev=false.

ahsrah7 commented:
> > > 1. There are no tests for disabled feature flag.
> > > 2. Could you add the test coverage for you changes?
> > 
> > 
> > 
> > * will the tests for the feature flag not become the tests once feature flag is removed?
> > * I have added the relevant test cases but  if you have any specific cases please mention
> 
> @ahsrah7 I don't see any test when dev=false.

@pankajjs  please let me know if I am getting this correctly. The existing flow where only super user  can edit the ERs already has test cases. For the changes I have added where a  normal user can also edit the ER until it is approved is  under FF so have added tests accordingly. For the case which you are mentioning dev=false should be the test cases for the already existing flow which are already present. If I am missing anything please let me know.

ahsrah7 commented:
> As I see in updateExtensionRequest, you have introduced dev flag. When dev is false, you are returning `res.boom.badRequest("Extension Request cannot be updated")` this. But you have not written test for this.

Yes I have introduced the dev flag but when dev flag is false the control does not reach the updateExtensionRequest at all. But thanks for raising this I notice that string type dev flag is always being type casted to true  irrespective of the value provided in the query params.

Achintya-Chatterjee commented on controllers/extensionRequests.js:
```diff
@@ -202,12 +202,23 @@ const getSelfExtensionRequests = async (req, res) => {
+      isDev &&
+      !req.userData?.roles.super_user &&
+      (extensionRequest.extensionRequestData.status === EXTENSION_REQUEST_STATUS.APPROVED ||
+        extensionRequest.extensionRequestData.status === EXTENSION_REQUEST_STATUS.DENIED)
+    ) {
+      return res.boom.badRequest("Extension Request cannot be updated");
+    }
```
`Extension Request cannot be updated` is a little bit generic. It would be beneficial to provide more context on why the request cannot be updated

ahsrah7 replied:
Got it. The message is updated to `Only pending extension request can be updated`. Please let me know if this is better.

vinit717 commented on middlewares/skipAuthorizeRolesWrapper.js:
Please write the test for the middleware

ahsrah7 replied:
Added the test cases for middleware. Please review

yesyash commented on controllers/extensionRequests.js:
```diff
@@ -202,12 +202,23 @@ const getSelfExtensionRequests = async (req, res) => {
+    if (
+      isDev &&
+      !req.userData?.roles.super_user &&
+      (extensionRequest.extensionRequestData.status === EXTENSION_REQUEST_STATUS.APPROVED ||
+        extensionRequest.extensionRequestData.status === EXTENSION_REQUEST_STATUS.DENIED)
```
- instead of checking for status not being approved (or) denied can we check for pending directly? so that if any new status is added in the future it does not slip through this check.

ahsrah7 replied:
Changed the condition to check for pending directly.

yesyash commented on middlewares/skipAuthorizeRolesWrapper.js:
- why are we doing this?

ahsrah7 replied:
Added this middleware to skip the roles check so so that normal user other than super user or app owner can edit a extension request after an extension request is created. As this has to be under feature flag a new middleware `skipAutorizaeRolesWrapper` is which skips the roles check when feature flag is true.

yesyash commented on routes/extensionRequests.js:
```diff
@@ -18,7 +19,7 @@ router.get("/:id", authenticate, authorizeRoles([SUPERUSER, APPOWNER]), extensio
   "/:id",
   authenticate,
-  authorizeRoles([SUPERUSER, APPOWNER]),
+  skipAuthorizeRolesUnderFF(authorizeRoles([SUPERUSER, APPOWNER])),
```
- why're we doing this?

ahsrah7 replied:
here the `skipAuthorizeRolesUnderFF` is invoked to skip the `authorizeRoles` middleware so that the end point is accessible to normal user for editing the extension request.

yesyash replied:
- do we plan to remove the `skipAuthorizeRolesUnderFF` & `authorizeRoles` middleware when removing the feature flag?

ahsrah7 replied:
Yes that is the plan. I have created an issues and added in the PR description under additional notes.
